### PR TITLE
For Noah to comment on Stewart's first swing at implementing the NN

### PR DESCRIPTION
**Pull request by sarenehan**

### CODE REVIEW

File: stewart_intro/generate_training_data.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+
+def get_initial_shape():
+    return load_data().QT.shape
+
+
+def get_target_normalization(variable):
+    data = load_data()
+    mean_by_vertical_slice = data[variable].mean(axis=(0, 2, 3)).values
+    std_by_vertical_slice = data[variable].mean(axis=(0, 2, 3)).values
```
Comment (nbren12) [inline stewart_intro/generate_training_data.py]: you should call `.std` here. It looks like you are normalizing by the mean. 

If the shape of the array is "(time, z, y, x)" then the normalization in the paper is `var.std(axis=(0,2,3)).mean(axis=0)`. The final mean across vertical levels is needed because the variance of QT is near 0 for the upper atmosphere (cold) vertical levels, and dividing by zero is bad.

File: stewart_intro/simple_nn.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+                for cell in batch_values
+            ]))
+    return (
+        Variable(torch.from_numpy(x_data), requires_grad=True),
+        Variable(torch.from_numpy(target), requires_grad=True),
+        Variable(torch.from_numpy(np.array(fqts_fslis)), requires_grad=True)
+    )
+
+
+def initialize_weights(data):
```
Comment (nbren12) [inline stewart_intro/simple_nn.py]: It's totally cool that you went for the functional interface, but for deeper networks with more parameters I have really liked using pytorch's `Module` class: https://pytorch.org/docs/stable/nn.html#torch.nn.Module.
It makes it really easy to keep track of parameters, save models to pickle files, etc. If you go that route, you would make `initialize_weights` and `predict` methods of an object inheriting from `torch.nn.Module`, and declare `w1` and `w2` in the `__init__` method.

File: stewart_intro/simple_nn.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    w2 = Variable((torch.rand(n_hidden_nodes, len(data.z) * 2) - .5) / 1000,
+                  requires_grad=True)
+    return w1, w2
+
+
+def predict(w1, w2, n_time_steps, x_data, fqts_fslis):
+    prediction = x_data[:, :-3]
+    for i in range(n_time_steps):
+        prediction = prediction + (dt * fqts_fslis[i])
+        to_predict_from = torch.transpose(torch.cat(
```
Comment (nbren12) [inline stewart_intro/simple_nn.py]: I have gotten better accuracy if I make the NN predict a source term (e.g. `prediction = prediction + dt * (output of nn)`) rather than `prediction = neural_network(prediction)`.

File: stewart_intro/simple_nn.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            (
+                torch.transpose(prediction, 0, 1),
+                torch.transpose(x_data[:, -3:], 0, 1),
+            )
+        ), 0, 1)
+        layer_one_out = F.relu(to_predict_from.matmul(w1))
+        prediction = layer_one_out.matmul(w2)
```
Comment (nbren12) [inline stewart_intro/simple_nn.py]: does this NN add a constant bias vector to the output of each layer?

File: stewart_intro/simple_nn.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    return prediction
+
+
+def squared_loss(y_hat, y):
+    return (1 / 2) * (y - y_hat).pow(2)
```
Comment (nbren12) [inline stewart_intro/simple_nn.py]: 1/2 not needed

File: stewart_intro/simple_nn.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+
+def compute_error(w1, w2, x_data, target, fqts_fslis):
+    if len(x_data.size()) == 1:
+        n = 1
+    else:
+        n = x_data.size()[0]
+    y_hat = predict(w1, w2, n_time_steps, x_data, fqts_fslis)
+    return ((1 / n) * squared_loss(
```
Comment (nbren12) [inline stewart_intro/simple_nn.py]: `((1/n) * x.sum())` is the same as `x.mean()`.

File: stewart_intro/simple_nn.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    n_batches = len(batches)
+    for epoch_num in range(n_epochs):
+        for idx, batch in enumerate(batches):
+            print(f'{idx} of {n_batches} for epoch {epoch_num}')
+            x_data, target, fqts_fslis = \
+                get_formatted_training_data_from_batch(data, batch)
+            err_est = compute_error(w1, w2, x_data, target, fqts_fslis)
+            print(float(err_est))
+            err_est.backward()
+            w1.data = w1.data - (training_rate * w1.grad.data)
```
Comment (nbren12) [inline stewart_intro/simple_nn.py]: Props on implementing this fixed learning rate gradient descent manually. very cool. In the future, you will probably want to use `Adam` or some fancier optimizer. Torch has a bunch of implementations here: https://pytorch.org/docs/stable/optim.html#how-to-use-an-optimizer.

File: stewart_intro/simple_nn.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+                np.concatenate([cell.QT.values, cell.SLI.values])
+                for cell in batch_values
+            ])
+        if i != n_time_steps:
+            fqts_fslis.append(np.array([
+                np.concatenate([cell.FQT.values, cell.FSLI.values])
+                for cell in batch_values
+            ]))
+    return (
+        Variable(torch.from_numpy(x_data), requires_grad=True),
```
Comment (nbren12) [inline stewart_intro/simple_nn.py]: Since v0.4 of pytorch, it is not necessary to wrap every tensor in `Variable`.  All tensors are now okay to use with autograd.

File: stewart_intro/simple_nn.py (hunk 4 above)
Comment (sarenehan) [reply stewart_intro/simple_nn.py]: It doesn't, but I could modify it by adding a constant weight vector to "to_predict_from" and "layer_one_output", and increasing the size of w1 and w2 by 1. I didn't get around to including bias terms for this V0.

File: stewart_intro/simple_nn.py (hunk 8 above)
Comment (sarenehan) [reply stewart_intro/simple_nn.py]: Good to know, thanks!

File: stewart_intro/simple_nn.py (hunk 2 above)
Comment (sarenehan) [reply stewart_intro/simple_nn.py]: Yes, the pytorch Module class definitely seems like the "right" way to do this -- i'll transition to that moving forward.

File: stewart_intro/generate_training_data.py (hunk 1 above)
Comment (sarenehan) [reply stewart_intro/generate_training_data.py]: Got it, thanks for clarifying. I'll update this.

File: stewart_intro/simple_nn.py (hunk 3 above)
Comment (sarenehan) [reply stewart_intro/simple_nn.py]: Interesting, so the NN is better at predicting the derivative than the next state. So for each time step is prediction = prediction + dt * (output of nn + fqt)?

File: stewart_intro/simple_nn.py (hunk 7 above)
Comment (sarenehan) [reply stewart_intro/simple_nn.py]: Yes, I am sure the accelerated methods will converge a lot faster. I will use those going forward.

File: stewart_intro/simple_nn.py (hunk 2 above)
Comment (nbren12) [reply stewart_intro/simple_nn.py]: ok sounds. good. let's see if your current code works before  you do too much refactoring.

File: stewart_intro/simple_nn.py (hunk 7 above)
Comment (sarenehan) [reply stewart_intro/simple_nn.py]: yeah i am getting convergence, it just takes a while. What batch size do you find works well?

Comment (nbren12) [reply stewart_intro/simple_nn.py]: maybe if you use a learning_rate of .001 it will be faster. Batch size shouldn't matter too much.

Comment (nbren12) [reply stewart_intro/simple_nn.py]: Adam really does make a big difference TBH
